src/Microsoft.IdentityModel.Abstractions/EventDetails.cs

```diff
@@ -0,0 +1,197 @@
+{
+    /// <summary>
+    /// Details of the telemetry event.
+    /// </summary>
+    public abstract class EventDetails
+    {
+        internal Dictionary<string, DataClassification> DataClassificationMapping = new Dictionary<string, DataClassification>();
+        internal Dictionary<string, string> PropertyValues = new Dictionary<string, string>();
+        internal Dictionary<string, bool> BoolPropertyValues = new Dictionary<string, bool>();
+        internal Dictionary<string, DateTime> DateTimePropertyValues = new Dictionary<string, DateTime>();
+        internal Dictionary<string, long> LongPropertyValues = new Dictionary<string, long>();
+        internal Dictionary<string, double> DoublePropertyValues = new Dictionary<string, double>();
+        internal Dictionary<string, Guid> GuidPropertyValues = new Dictionary<string, Guid>();
+        internal Dictionary<string, DictionaryTypeEnum> PropertiesKeys = new Dictionary<string, DictionaryTypeEnum>();
+
+        /// <summary>
+        /// Name of the telemetry event, should be unique between events.
+        /// </summary>
+        public abstract string Name { get; set; }
+
+        /// <summary>
+        /// Properties which describe the event.
+        /// </summary>
+        public virtual Dictionary<string, object> Properties
+        {
+            get
+            {
+                var currentBag = new Dictionary<string, object>();
+                foreach (var kvp in PropertiesKeys)
+                {
+                    switch (kvp.Value)
+                    {
+                        case DictionaryTypeEnum.String:
+                            currentBag.Add(kvp.Key, PropertyValues[kvp.Key]);
+                            break;
+                        case DictionaryTypeEnum.Long:
+                            currentBag.Add(kvp.Key, LongPropertyValues[kvp.Key]);
+                            break;
+                        case DictionaryTypeEnum.Double:
+                            currentBag.Add(kvp.Key, DoublePropertyValues[kvp.Key]);
+                            break;
+                        case DictionaryTypeEnum.Bool:
+                            currentBag.Add(kvp.Key, BoolPropertyValues[kvp.Key]);
+                            break;
+                        case DictionaryTypeEnum.DateTime:
+                            currentBag.Add(kvp.Key, DateTimePropertyValues[kvp.Key]);
+                            break;
+                        case DictionaryTypeEnum.Guid:
+                            currentBag.Add(kvp.Key, GuidPropertyValues[kvp.Key]);
+                            break;
+                    }
+                }
+                return currentBag;
+            }
+        }
+
+        /// <summary>
+        /// Gets the <see cref="DataClassification"/> mapping to keys in <see cref="Properties"/>.
+        /// </summary>
+        public virtual Dictionary<string, DataClassification> PropertyDataClassification
```

Comment:
When would I override all these virtual methods? Would it make sense to keep them as normal method (non-virtual) and add the virtual property later (as a non-breaking change) ?

Reply:
might do so to be thread-safe, or otherwise extend the event details.